lib/srv/db/mongodb/protocol/message.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package protocol
+
+import (
+	"bytes"
+	"context"
+	"io"
+
+	"go.mongodb.org/mongo-driver/x/mongo/driver"
+	"go.mongodb.org/mongo-driver/x/mongo/driver/wiremessage"
+
+	"github.com/gravitational/trace"
+)
+
+// Message defines common interface for MongoDB wire protocol messages.
+type Message interface {
+	// GetHeader returns the wire message header.
+	GetHeader() MessageHeader
+	// GetBytes returns raw wire message bytes.
+	GetBytes() []byte
+	// ToWire coverts the message to wire bytes format.
+	ToWire(responseTo int32) []byte
+}
+
+// ReadMessage reads the next MongoDB wire protocol message from the reader.
+func ReadMessage(reader io.Reader) (Message, error) {
+	header, payload, err := readHeaderAndPayload(reader)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	// Parse the message body.
+	switch header.OpCode {
+	case wiremessage.OpMsg:
+		return readOpMsg(*header, payload)
+	case wiremessage.OpQuery:
+		return readOpQuery(*header, payload)
+	case wiremessage.OpReply:
+		return readOpReply(*header, payload)
+	default:
+		return &MessageUnknown{
+			Header: *header,
+			raw:    payload,
+		}, nil
+	}
+}
+
+// ReadServerMessage reads wire protocol message from the MongoDB server connection.
+func ReadServerMessage(ctx context.Context, conn driver.Connection) (Message, error) {
+	var wm []byte
+	wm, err := conn.ReadWireMessage(ctx, wm)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return ReadMessage(bytes.NewReader(wm))
+}
+
+func readHeaderAndPayload(reader io.Reader) (*MessageHeader, []byte, error) {
+	// First read message header which is 16 bytes.
+	var header [headerSizeBytes]byte
+	if _, err := io.ReadFull(reader, header[:]); err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+	length, requestID, responseTo, opCode, _, ok := wiremessage.ReadHeader(header[:])
+	if !ok {
+		return nil, nil, trace.BadParameter("failed to read message header %v", header)
+	}
+	// Then read the entire message body.
+	payload := make([]byte, length-headerSizeBytes)
+	if _, err := io.ReadFull(reader, payload); err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+	return &MessageHeader{
+		MessageLength: length,
+		RequestID:     requestID,
+		ResponseTo:    responseTo,
+		OpCode:        opCode,
+		bytes:         header,
+	}, payload, nil
```

Comment:
Hmm, I'm not sure we can, for a couple of reasons:
* Not sure how using `io.Reader` helps since the ReadHeader methods expects a byte slice.
* If you meant `io.ReadAll`, then I don't think we can use it since we need to read exactly one wire message.

The way this works is:
1. We read 16 byte header which includes the total wire message length.
2. Then we read exactly the remaining message payload of this length.

This guarantees we will read a single wire message from the connection.

Reply:
Yes, I have meant `io.ReadAll` 

>his guarantees we will read a single wire message from the connection.

The reader contains bytes already readed by`wm, err := conn.ReadWireMessage(ctx, wm)` when ReadWireMessage takes into account the message length thus I thought that this is already guaranteed.
```
func ReadServerMessage(ctx context.Context, conn driver.Connection) (Message, error) {
	var wm []byte
	wm, err := conn.ReadWireMessage(ctx, wm)
	if err != nil {
		return nil, trace.Wrap(err)
	}
	return ReadMessage(bytes.NewReader(wm))
```